### PR TITLE
Calcium neuron model

### PR DESCRIPTION
**Pull request by james4424**

### CODE REVIEW

**Comment by james4424** [comment]
2 multi-compartment neuron models with calcium spikes added, 1 modeled using first order kinetics and the other using threshold triggered fixed waveform.

**Comment by tammoippen** [comment]
Hello @james4424 ,

The TravisCI complains about the formatting not complying to our [coding style guidelines](https://nest.github.io/nest-simulator/coding_guidelines_c++). Please have a look at the [tooling](https://nest.github.io/nest-simulator/coding_guidelines_c++#tooling) section and use `clang-format-3.6` to fix the formatting. Also, there is script, that fixes the formatting (provided that `clang-format-3.6` is installed) - calling those two lines should fix the formatting:

    ./extras/format_all_c_c++_files.sh models
    ./extras/format_all_c_c++_files.sh nestkernel

If you want some parts of the code to not be formatted, surround them with:

    // clang-format off
    ....
    // clang-format on

Best,
Tammo

**Comment by james4424** [comment]
Hi Tammo,
while trying to follow the instructions in https://nest.github.io/nest-simulator/coding_guidelines_c++#tooling
I encountered error while running the below:
# To retrieve the archive signature:
wget -O - http://llvm.org/apt/llvm-snapshot.gpg.key | sudo apt-key add -
sudo sh -c 'echo -n "deb http://llvm.org/apt/trusty/ llvm-toolchain-trusty-3.6 main" >> /etc/apt/sources.list'
sudo apt-get update
sudo apt-get install clang-format-3.6When I run sudo apt-get update, the error is :


W: Failed to fetch http://llvm.org/apt/trusty/dists/llvm-toolchain-trusty-3.6/Release  Unable to find expected entry 'maindeb/binary-amd64/Packages' in Release file (Wrong sources.list entry or malformed file)
E: Some index files failed to download. They have been ignored, or old ones used instead.


And if I run sudo apt-get install clang-format-3.6
then I would encounter the below error:

Some packages could not be installed. This may mean that you haverequested an impossible situation or if you are using the unstabledistribution that some required packages have not yet been createdor been moved out of Incoming.The following information may help to resolve the situation:
The following packages have unmet dependencies: clang-format-3.6 : Depends: libllvm3.6 but it is not going to be installedE: Unable to correct problems, you have held broken packages.


I am not sure how this can be resolved.
I am running Ubuntu 12.04 LTS.
Please advise.
best regards,yansong
Date: Wed, 2 Dec 2015 00:45:03 -0800

**Comment by tammoippen** [comment]
Hi,

You are running Ubuntu Precise 12.04, and the instructions are for Trusty 14.04. Try the following lines for Precise:

    # You added the gpg key already, so maybe this first line is not necessary
    # wget -O - http://llvm.org/apt/llvm-snapshot.gpg.key | sudo apt-key add -
    sudo sh -c 'echo -n "deb http://llvm.org/apt/trusty/ llvm-toolchain-trusty-3.6 main"  >> /etc/apt/sources.list'
    sudo apt-get update
    sudo apt-get install clang-format-3.6

Maybe you have to remove the line `deb http://llvm.org/apt/trusty/ llvm-toolchain-trusty-3.6 main` from `/etc/apt/sources.list`. 

Best,
Tammo

**Comment by james4424** [comment]
Hi Tammo,
you do mean for precise 12.04 :
sudo sh -c 'echo -n "deb http://llvm.org/apt/precise/ llvm-toolchain-precise-3.6 main"  >> /etc/apt/sources.list'instead of sudo sh -c 'echo -n "deb http://llvm.org/apt/trusty/ llvm-toolchain-trusty-3.6 main"  >> /etc/apt/sources.list'
right?I have also removed deb http://llvm.org/apt/trusty/ llvm-toolchain-trusty-3.6 mainand checked that it is replaced withdeb http://llvm.org/apt/precise/ llvm-toolchain-precise-3.6 main
However after running  sudo apt-get update, i encounter the below errors:......W: Failed to fetch http://extras.ubuntu.com/ubuntu/dists/precise/main/i18n/Translation-en  Unable to connect to extras.ubuntu.com:http:

E: Some index files failed to download. They have been ignored, or old ones used instead.

And when I run sudo apt-get install clang-format-3.6, I get:
Reading package lists... DoneBuilding dependency tree       Reading state information... DoneSome packages could not be installed. This may mean that you haverequested an impossible situation or if you are using the unstabledistribution that some required packages have not yet been createdor been moved out of Incoming.The following information may help to resolve the situation:
The following packages have unmet dependencies: clang-format-3.6 : Depends: libllvm3.6 but it is not going to be installed                    Depends: libstdc++6 (>= 4.9) but 4.6.3-1ubuntu5 is to be installedE: Unable to correct problems, you have held broken packages.

Hence, could you please advise how I may proceed. Would I need to upgrade to trusty 14.04?
Please advise.
best regards,yansong

Date: Wed, 2 Dec 2015 11:02:51 -0800

**Comment by tammoippen** [comment]
Sorry for the hassle. I tried installing `clang-format-3.6` on Precise on my own (and will update the documentation accordingly). Here is how you do it:
* You need the gpg key (you already done this):

        wget -O - http://llvm.org/apt/llvm-snapshot.gpg.key | sudo apt-key add -
* You need two new repository entries in `/etc/apt/sources.list`:

        sudo sh -c 'echo "\ndeb http://llvm.org/apt/precise/ llvm-toolchain-precise-3.6 main" >> /etc/apt/sources.list'
        sudo sh -c 'echo "deb http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu precise main" >> /etc/apt/sources.list'
  afterwards make sure, each entry is on its own line in the file.
* Then you can install `clang-format`:

        sudo apt-get update
        sudo apt-get install libstdc++6 libllvm3.6 clang-format-3.6
  It will complain, that the package cannot be verified, because you do not have the gpg key for _launchpad.net_ (you can get it following these [instructions](https://askubuntu.com/questions/308760/w-gpg-error-http-ppa-launchpad-net-precise-release-the-following-signatures/358424#358424)), but it still works.

I hope this helps!

Best,
Tammo

**Comment by james4424** [comment]
Hi Tammo,
I followed your instruction and managed to install clang
I also managed to install cppcheck.
however when i try to install vera++ by running 
 sudo apt-get install vera++
it does not seem to install the right software, as running vera++ version does not give me anything.
I tried installing from source (by downloading the tar.gz file), but when i followed the instructions for installation from its webpage
https://bitbucket.org/verateam/vera/wiki/developers/Building.md
cd vera
mkdir build
cd build
cmake ..
make -j
make install
specifically, when i run cmake ..
i get the below error
-- Could NOT find Lua51 (missing:  LUA_LIBRARIES LUA_INCLUDE_DIR) -- Looking for Luabind...-- Could NOT find Luabind (missing:  LUABIND_LIBRARIES LUABIND_INCLUDE_DIR) CMake Error at src/lua.cmake:10 (message):  Could NOT find Lua.  Turn VERA_USE_SYSTEM_LUA to OFF to build it with  vera++.Call Stack (most recent call first):  src/CMakeLists.txt:26 (include)

-- Configuring incomplete, errors occurred!
I suppose again it is some dependencies
What should I do now?
Please advise.
warmest regards,yansongDate: Fri, 4 Dec 2015 02:05:23 -0800

**Comment by tammoippen** [comment]
Can you please re-run:

    ./extras/format_all_c_c++_files.sh models
Sometimes, clang-format gets not all format changes in the first run.

**Comment by james4424** [comment]
Hi Tammo,

done!

it seems that now I have six out 8 test ok.

For the other 2, the common error is the below:

  ERROR: test_FixedMultiCmptNeu
(nest.tests.test_fixed_multi_cmpt_neu.FixedMultiCmptNeuTestCase)

      raise _kernel.NESTError("{0} in {1}{2}".format(errorname,
commandname, message))

  NESTError: UnknownModelName in GetDefaults_l:
/iaf_cond_alpha_mc_fixedca is not a known model name. Please check the
modeldict for a list of available models. A frequent cause for this
error is that NEST was compiled without the GNU Scientific Library,
which is required for the conductance-based neuron models.

  ERROR: test_KineticsMultiCmptNeu
(nest.tests.test_kinetics_multi_cmpt_neu.KineticsMultiCmptNeuTestCase)

      raise _kernel.NESTError("{0} in {1}{2}".format(errorname,
commandname, message))

  NESTError: UnknownModelName in GetDefaults_l:
/iaf_cond_alpha_mc_kinetics is not a known model name. Please check
the modeldict for a list of available models. A frequent cause for
this error is that NEST was compiled without the GNU Scientific
Library, which is required for the conductance-based neuron models.

  FAILED (SKIP=13, errors=2)


any idea how to fix this?

best regards,
yansong

File: pynest/nest/tests/test_fixed_multi_cmpt_neu.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+import nest
+from scipy import *
+from scipy.integrate import ode
+import unittest
+
+HAVE_GSL = nest.sli_func("statusdict/have_gsl ::")
+
+@unittest.skipIf(not HAVE_GSL, 'GSL is not available')
+
+class mc_neuron_ode():
```
Comment (tammoippen) [inline pynest/nest/tests/test_fixed_multi_cmpt_neu.py]: Please also have the `@nest.check_stack` in front of the class definition.

File: pynest/nest/tests/test_fixed_multi_cmpt_neu.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+		if self.reftmp > 0:
+			self.reftmp -=1
+			self.catmp = self.ca[self.refca - self.reftmp]
+		else:
+			if self.u[2] >= self.thca:
+				self.reftmp = self.refca
+				self.catmp = self.ca[0]
+		
+
+class FixedMultiCmptNeuTestCase(unittest.TestCase):
```
Comment (tammoippen) [inline pynest/nest/tests/test_fixed_multi_cmpt_neu.py]: This `TestCase` also needs the decorators:

    @nest.check_stack
    @unittest.skipIf(not HAVE_GSL, 'GSL is not available')

File: pynest/nest/tests/test_kinetics_multi_cmpt_neu.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+import nest
+from scipy import *
+from scipy.integrate import ode
+import unittest
+
+HAVE_GSL = nest.sli_func("statusdict/have_gsl ::")
+
+@unittest.skipIf(not HAVE_GSL, 'GSL is not available')
+
+class mc_neuron_ode():
```
Comment (tammoippen) [inline pynest/nest/tests/test_kinetics_multi_cmpt_neu.py]: This `TestCase` also needs the decorators:

    @nest.check_stack

Comment (tammoippen) [inline pynest/nest/tests/test_kinetics_multi_cmpt_neu.py]: See above.

File: pynest/nest/tests/test_kinetics_multi_cmpt_neu.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+		self.r.y[7] += (G_exc[1,0]*np.exp(1)/self.Tau_synE[1,0])
+		self.r.y[9] += (G_inh[1,0]*np.exp(1)/self.Tau_synI[1,0])
+		self.r.y[3] += (G_exc[2,0]*np.exp(1)/self.Tau_synE[2,0])
+		self.r.y[5] += (G_inh[2,0]*np.exp(1)/self.Tau_synI[2,0])
+
+		self.r.integrate(self.r.t+self.resolution)
+		self.u = self.r.y
+
+
+class KineticsMultiCmptNeuTestCase(unittest.TestCase):
```
Comment (tammoippen) [inline pynest/nest/tests/test_kinetics_multi_cmpt_neu.py]: This `TestCase` also needs the decorators:

    @nest.check_stack
    @unittest.skipIf(not HAVE_GSL, 'GSL is not available')

Comment (james4424) [inline pynest/nest/tests/test_kinetics_multi_cmpt_neu.py]: Hi Tammo,

the error persists even after making the corrections

yansong

**Comment by james4424** [comment]
Now that we are moving into the review stage, I would like to take the chance to propose a few reviewers namely, @mhelias and @heplesser, thanks!

**Comment by jougs** [comment]
@james4424: can you please resolve the conflicts mentioned above? Thanks!

**Comment by james4424** [comment]
Hi Jochen,

isn't it already resolved?

best regards,
yansong

**Comment by jougs** [comment]
The status message at the end of the webpage for the pull request says "This branch has conflicts that must be resolved". Usually this means that a manual merge from master to your branch is necessary in order to make automatic merging possible for the pull request.

**Comment by heplesser** [comment]
@james4424 I am sorry for the long delay. I will try to get around to reviewing the code later this week. 

The merge conflict is due to the fact that we recently changed the internal structure of the simulator kernel (see #205). This will require some minor changes in the code for your model, mainly related to `get_network()` calls. Could you update your NEST repo to the current state? The looking at how `iaf_cond_alpha_mc` code changed in the #205 commit should give you a very good guideline of how to update your model. I'd be very happy to help if you run into any problems.

@tammoippen Do we have a guideline on how to adapt neuron models to the new kernel structure somewhere?

**Comment by tammoippen** [comment]
No, we do not have a guideline yet. The changes are minimal:

* Instead of including `network.h` include `kernel_manager.h`.
* For sending a spike instead of `network()->send( *this, se, lag );` use `kernel().event_delivery_manager.send( *this, se, lag );`
* To get the slice origin instead of `network()->get_slice_origin()` use `kernel().simulation_manager.get_slice_origin()`

**Comment by james4424** [comment]
Hi, I have to create another branch to work on this for the following reasons:

it has been approximately 2 months since I last committed a tested and working version. since then, i have added stuff to the nest on my laptop which there is no need to merge with master. in the 2 months, i notice changes have been made to the master copy such that what was working before is now not. 

initially, i thought i could just modified the neuron model cpp files and then merge it with this branch and re-do the pull request. however i realize, by doing so, i would break my own nest ( on my laptop). while to  use the newest master copy as my current and working nest version would involve having to re-add other stuff into it, and again, i would have to then think of a way to avoid pushing them into git.

hence, the best option (that i know of) is simply to re-create another branch, and work with the calcium models in the new branch. 

i have done that, and i realize that there are errors during make : 
error: ‘Scheduler’ has not been declared
apparently, there are changes made not mentioned above. i tried looking for it, but i am not sure if i should now instead call kernel().connection_builder_manager.get_min_delay() or something else.

**Comment by james4424** [comment]
@heplesser @tammoippen the other branch has passed all checks, please take a look and merge it with master if there are no problems

File: models/iaf_cond_alpha_mc_fixedca.h [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+Meffin, H., Burkitt, A. N., & Grayden, D. B. (2004). An analytical
+model for the large, fluctuating synaptic conductance state typical of
+neocortical neurons in vivo. J.  Comput. Neurosci., 16, 159–175.
+
+Bernander, O ., Douglas, R. J., Martin, K. A. C., & Koch, C. (1991).
+Synaptic background activity influences spatiotemporal integration in
+single pyramidal cells.  Proc. Natl. Acad. Sci. USA, 88(24),
+11569–11573.
+
+Chua, Y., Morrison, A., & Moritz, H. (2015).
```
Comment (mhelias) [inline models/iaf_cond_alpha_mc_fixedca.h]: last name is Helias

File: models/iaf_cond_alpha_mc_fixedca.h [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+integrate-and-fire
+neuron model with calcium spike at distal compartment modeled using threshold triggered fixed
+waveform.
+
+Description:
+THIS MODEL IS A PROTOTYPE FOR ILLUSTRATION PURPOSES. IT IS NOT YET
+FULLY TESTED. USE AT YOUR OWN PERIL!
+
+iaf_cond_alpha_mc_fixedca is an implementation of a multi-compartment spiking
+neuron using IAF dynamics with conductance-based synapses. It serves
```
Comment (mhelias) [inline models/iaf_cond_alpha_mc_fixedca.h]: add a comment here that the implementation is described in reference 4 below

**Comment by mhelias** [comment]
The code looks fine to me. I have left two comments regarding the documentation above.

**Comment by tammoippen** [comment]
This PR is replaced by #241.

**Pull request closed without merging**